### ISSUE TITLE
Skip currently running task and proceed to next

### ISSUE THREAD

**Issue by Avinm** [jessetane/queue#48]
Is there a way one can get the queue to ignore the current running/pending task and proceed to the next task?

**Comment by jessetane** [comment]
well there is the `timeout` property?

**Comment by Avinm** [comment]
Yeah, but what if one needs to cancel it based on a uswr interaction?

Regards,
Avin
________________________________

**Comment by jessetane** [comment]
well if you actually want to cancel some async io (like an http request), wouldn't it be more appropriate to handle that in the job itself?

``` javascript
function myjob (cb) {
  ui.on('cancel', oncancel)
  var req = request('http://some.url', (err, res, body) => {
    ui.removeListener('cancel', oncancel)
    if (err) return cb(err)
    // do something with data
    cb()
  })
  function oncancel () {
    req.abort()
  }
}
```

**Comment by Avinm** [comment]
This will cancel all the jobs of this kind when a cancel event is raised. I'd like to cancel only the running job in the queue

**Comment by jessetane** [comment]
how would it cancel anything other than running jobs?

**Comment by Avinm** [comment]
Uh oh, my bad. I didn't notice that the listener is attached only after the invoke. But even so, I'd prefer if myJob remained unaware of the ui and that the queue takes care of it

**Comment by jessetane** [comment]
how would queue know how to cancel an http request?

**Comment by Avinm** [comment]
I'd not expect the queue to cancel the request, but ignore that job altogether. Is this possible?

**Comment by jessetane** [comment]
i think that encourages bad behavior (ui says cancel, but that's not what actually happens - io, timers, etc are left running) but if you must do this i'm pretty sure you would have to change the implementation. if it seems easy send a patch with a test and we can discuss more.

**Comment by Avinm** [comment]
UI is just one example. This could also be useful if one wants to move a running job to another queue as well. I will look around to see what I can do.

**Comment by jessetane** [comment]
fair enough. i suspect we'd need to start keeping track of all running jobs via an instance property but haven't thought about it too hard..

**Comment by crowmagnumb** [comment]
Here is a hack of sorts I came up with to have cancelable jobs in the queue.

```
var queue = require("queue");

process.on("unhandledRejection", error => {
    // just avoid node from barfing and displaying error.
});

var q = queue({
    timeout: 3000,
    autostart: true,
    concurrency: process.argv.length > 2 ? process.argv[2] : 2
});

function makeTask(id, duration) {
    const fn = function() {
        let timer;
        fn.promise = new Promise((resolve, reject) => {
            fn.reject = reject;

            timer = setTimeout(() => {
                resolve(id * id);
            }, duration);
        });
        fn.promise.finally(() => {
            clearTimeout(timer);
        });
        return fn.promise;
    };

    fn.cancel = function() {
        if (fn.reject) {
            fn.reject("Canceled");
        }
    };
    fn.id = id;
    return fn;
}

q.on("timeout", function(next, job) {
    console.log(`job [${job.id}] timed out.`);
    next();
});

q.on("success", function(result, job) {
    console.log(`job [${job.id}] completed with result [${result}].`);
});

q.on("error", function(err, job) {
    console.log(`job [${job.id}] failed with error [${err}].`);
});

q.on("start", function(job) {
    console.log(`job [${job.id}] started.`);
    // console.log(job.toString());
    // console.log(q.results);
    if (job.id === 2) {
        setTimeout(() => {
            job.cancel();
        }, 1000);
    }
});

q.push(makeTask(1, 1000));
q.push(makeTask(2, 2000));
q.push(makeTask(3, 1500));
q.push(makeTask(4, 5000));
q.push(makeTask(5, 2500));
```

**Comment by jessetane** [comment]
Why do you call this a hack? It seems pretty reasonable to me.

Except maybe this bit, which is unnecessary and potentially dangerous:
```
process.on("unhandledRejection", error => {
    // just avoid node from barfing and displaying error.
});
```

**Comment by jessetane** [comment]
@crowmagnumb this isn't aimed at you specifically so please forgive the rant, but I've been bitten so many times by stuff like this (in particular incorrect usage of try/catch) I just feel compelled to yell about it a little. If you don't know what caused (or how to handle) an error, the best policy is almost always to let your program crash. After it crashes, you'll have plenty of time to figure out whatever the actual problem was and to fix it properly. Running a service in production and can't afford the downtime? Let init restart your service for you, and remember to review your logs from time to time. Writing a library? Please do not attempt to handle errors caused by bad user input, either pass an error back via whatever the paradigm is or crash - if you don't, chances are the user will eventually end up with some extremely difficult to track down bugs.

**Comment by crowmagnumb** [comment]
Didn't mean to leave that in. I was getting that when I was trying different things and I couldn't figure out the source of it and I just wanted to concentrate on the queue at hand. I'm going to edit it out above (doing it this way for future browsers to find the clean first one rather than have to dig for the second).

Appreciate the feedback concern on that, didn't need the lecture that assumed I was unaware of the implications. Please refrain from slipping into man-splaining. ;) Basically you would have been fine with the first comment alone.  :)

As for calling it a hack? Well loose term I guess.

**Comment by jessetane** [comment]
As I said, the rant wasn’t aimed at you, your program clearly shows you know what you’re doing. The lecture was more for others who might stumble upon the thread later. In any case, you’re right I shouldn’t have written it the way I did. I’m sorry and thanks for updating your example.
